### ISSUE TITLE
Why don't you use e4e in the first stage Inversion?

### ISSUE THREAD

**Issue by lastapple** [danielroich/PTI#9]
Great work! 
I notice that there is a "first_inv_type" option in hyperparameters.py. You must have tried using e4e in the first stage. Can you tell us  why you choose to use the original projector in StyleGAN2 instead of e4e? Thanks!

**Comment by danielroich** [comment]
Hey @lastapple ,
Thanks!

The paper of "e4e" found out that there is an inherent tradeoff between editability and id-preservation. 
This means that the closer you get to *W* space the edits are much more efficient and behave as expected without artifacts.
While the more you deviate from *W* and move to *W+* you will be awarded with better reconstruction and the image will be more similar to the target image.
From this, you can understand that in order to edit real images. I am talking about celebrities, your family and much more you must invert to *W+* Space. (Disclaimer: There will be some images that could be inverted pretty good close to *W*, but they are not much)

e4e successfully tries to take advantage of this fact by inverting the image to a "sweet spot" in-between *W* and *W+*.

In theory (and practically, by changing first_inv_type value to "e4e") you could use e4e in the first step. But it would result in less editable code. The Id preservation in PTI will always be the same, near-perfect for most cases, regardless of the first inversion type.

It will only affect the editability of the resulted latent code. By inverting the image to *W* instead of *W+* or any other space.,
PTI accomplishes both near-perfect inversions **and** the best editability. Thus breaking the trade-off found in e4e. If you change the first inversion to e4e, you will get less editable code.

I think once you examine the comparisons we have made in the paper. Which can be found [here](https://www.google.com/search?q=pivotal+tuning+for+latent-based+editing+of+real+images&oq=pivo&aqs=chrome.0.69i59j0j69i57j69i59l2j69i61j69i60l2.701j0j7&sourceid=chrome&ie=UTF-8) or at the github page you will see what I am talking about.

Hope it helps.
If not ,feel free to say so and I will write even longer answer,
Daniel

**Comment by danielroich** [comment]
You can see in `notebooks/inference_playground.ipynb` that when using StyleCLIP, I am configuring PTI to use e4e instead of *W* inversion. This is done because StyleCLIP was trained on e4e inversion. And because of that, StyleCLIP works better on latent with e4e editability - it has bias towards e4e latents.

**Comment by lastapple** [comment]
Oh it‘s’ a bit embarrassing, I didn't expect that you would reply so quickly and patiently. Later I see the ablation study in Fig.16 where you have explained this. The results is very interesting and I will read more carefully.
